rfcs/sexp.md

```diff
@@ -0,0 +1,104 @@
+# Add a `Sexp` module to the standard library
+
+I propose to add a `Sexp` module to the standard library.
+The goal is twofold:
+
+1. Provide a common type definition that can be shared
+   among libraries that deal with s-expressions.
+
+2. Promote the use of s-expressions as a lightweight
+   human-readable serialization format.
+
+The module would include at least the following definitions:
+
+```
+type t =
+  | Atom of string
+  | List of t list
+
+val equal: t -> t -> bool
+
+val compare: t -> t -> int
+
+val pp: Format.formatter -> t -> unit
+
+val to_string: t -> string
+
+val of_string: string -> t option
+```
+
+**Question:** The above functions will pretty-print their arguments.  Should
+"compact" variants (minimum amount of whitespace) also be included?
+
+## Support in other modules of the standard library
+
+In addition, functions
+
+- `to_sexp : t -> Sexp.t`
+- `of_sexp_opt: Sexp.t -> t option`
+- `of_sexp: Sexp.t -> t`
+
+would be added to all existing modules of the stdlib for which it makes sense:
+
+- `Unit`
+- `Char`
+- `Uchar`
+- `Bool`
+- `Int`
+- `Int32`
+- `Int64`
+- `Float`
+- `String`
+- `Bytes`
+
+Container types would have suitable functions added to them:
+
+- `Lazy`, `Seq`
+
+```
+val to_sexp: ('a -> Sexp.t) -> 'a t -> Sexp.t (* forces the argument *)
+val of_sexp_opt: (Sexp.t -> 'a option) -> Sexp.t -> 'a t option
```

Comment:
Promising a `'a t option` means that you have to consume the whole serialized sequence to know in advance if one of the elements is going to fail. Maybe a comment saying `(* returns a strict collection *)` could clarify this, and it would also make sense on `of_sexp` for which either strict and lazy implementations are possible (if an element is invalid, do we get an exception now or only when trying to access it?).

(To represent this nicely one would actually need an `('a, 'b) Seq.t` that is a lazy stream of `'a` followed by a `'b`, with normal streams being `('a, unit) Seq.t`.)

Reply:
Good point. I would expect `of_sexp` to be lazy by default for these types (this is what I had in mind when I wrote it, I didn't realize the point about the `option`-returning one).

Reply:
But `of_sexp_opt` and `of_sexp` should behave in the same way, shouldn't them?